### PR TITLE
IGNITE-14306 Hint ENFORCE JOIN ORDER support

### PR DESCRIPTION
**Pull request by ygerzhedovich**

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/Hints.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        return hints.stream().anyMatch(this::is);
+    }
+
+    /**
+     * Return {@code true} if the the hint presents in provided hint list and options are presents.
+     *
+     * @param hints Hints.
+     * @return {@code true} If found the hint with rules for disabling, {@code false} otherwise.
+     */
+    public boolean presentWithParams(List<RelHint> hints) {
```
Comment (lowka) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/Hints.java:84]: I think it would be better to convert Hints to a close hierarchy of classes and have a single method that transforms a collection of RelHint s to Hints supported by ignite.
It would allow this would make `present` method redundant and would allow to keep only a single variant of `toHint`.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java [hunk 2]
```diff
@@ -1318,4 +1344,49 @@ public Publisher<BinaryRow> lookup(int partId, HybridTimestamp timestamp, Cluste
...
+     * Listener which keeps information about attempts of applying optimization rules.
+     */
+    public class RuleAttemptListener extends RelOptListenerWrapper {
+        Set<String> attemptedRules = new HashSet<>();
+
+        @Override
+        public void ruleAttempted(final RuleAttemptedEvent event) {
+            if (event.isBefore()) {
+                attemptedRules.add(event.getRuleCall().getRule().toString());
+                System.out.println(event.getRuleCall().getRule().toString());
```
Comment (lowka) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java:1380]: Is this print to stdout necessary?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/Commons.java [hunk 3]
```diff
@@ -142,8 +142,9 @@ public final class Commons {
                     .withExpand(false)
                     .withHintStrategyTable(
                             HintStrategyTable.builder()
-                                    .hintStrategy("DISABLE_RULE", (hint, rel) -> true)
-                                    .hintStrategy("EXPAND_DISTINCT_AGG", (hint, rel) -> rel instanceof Aggregate)
+                                    .hintStrategy(Hints.ENFORCE_JOIN_ORDER.name(), (hint, rel) -> true)
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/Commons.java:145]: why did we want to hold predicate rule configuration as lambdas ? i.e. : rel instanceof Aggregate ? 
check for example : org.apache.calcite.rel.hint.HintPredicates seems it`s more clear and seems it`s @lowka is talking about.
Additionally why new rule is propageted for all rels ? Not for : **rel instanceof Join** ?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java [hunk 4]
```diff
@@ -97,8 +98,10 @@ public static IgniteRel optimize(SqlNode sqlNode, IgnitePlanner planner) {
 
             joinSizeFinder.visit(rel);
 
-            if (joinSizeFinder.sizeOfBiggestJoin() > MAX_SIZE_OF_JOIN_TO_OPTIMIZE) {
-                Set<String> disabledRules = new HashSet<>(HintUtils.disabledRules(root.hints));
+            boolean amountOfJoinsAreBig = joinSizeFinder.sizeOfBiggestJoin() > MAX_SIZE_OF_JOIN_TO_OPTIMIZE;
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java:101]: Seems more clear to move it into separate function isn`t it ?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/Commons.java (hunk 3 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/Commons.java]: good catch, thanks. Fixed. But it's not the same which talk @lowka about

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java (hunk 4 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java]: agree

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java [hunk 5]
```diff
@@ -1318,4 +1344,48 @@ public Publisher<BinaryRow> lookup(int partId, HybridTimestamp timestamp, Cluste
...
+            if (event.isBefore()) {
+                attemptedRules.add(event.getRuleCall().getRule().toString());
+            }
+        }
+
+        public boolean isApplied(String rule) {
+            return attemptedRules.stream().anyMatch(r -> r.contains(rule));
+        }
+
+        public void resetStatistics() {
```
Comment (zstan) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java:1387]: statistic - confusing, probably simple - **reset** is enough ?

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/JoinCommutePlannerTest.java [hunk 6]
```diff
@@ -76,13 +79,44 @@ public static void init() {
...
+    public void testEnforceJoinOrderHint() throws Exception {
+        String sqlJoinCommuteWithNoHint = "SELECT COUNT(*) FROM SMALL s, HUGE h, HUGE h1 WHERE h.id = s.id and h1.id=s.id";
+        String sqlJoinCommuteWithHint =
+                "SELECT /*+ ENFORCE_JOIN_ORDER */ COUNT(*) FROM SMALL s, HUGE h, HUGE h1 WHERE h.id = s.id and h1.id=s.id";
+        String sqlJoinCommuteOuterWithNoHint = "SELECT COUNT(*) FROM SMALL s RIGHT JOIN HUGE h on h.id = s.id";
+        String sqlJoinCommuteOuterWithHint = "SELECT /*+ ENFORCE_JOIN_ORDER */ COUNT(*) FROM SMALL s RIGHT JOIN HUGE h on h.id = s.id";
+
+        RuleAttemptListener listener = new RuleAttemptListener();
+
+        physicalPlan(sqlJoinCommuteWithNoHint, publicSchema, listener);
```
Comment (zstan) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/JoinCommutePlannerTest.java:100]: i suggest to use real rule names instead of hadcoded, one change for rule namings will broke huge amount of tests :
JoinConditionPushRuleConfig rule = JoinConditionPushRuleConfig.DEFAULT;
String ruleName = rule.toRule().toString();

**Review by zstan — APPROVED** [review]
i fill some comments, plz check them, overall looks good

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/WorkDirectoryExtensionTest.java [hunk 7]
```diff
@@ -380,7 +380,7 @@ void test() {
     /**
      * Tests {@code WorkDirectoryExtension} lifecycle works correctly on a test class with all test methods being disabled.
      *
-     * @see <a href="https://issues.apache.org/jira/browse/IGNITE-15799">IGNITE-15799</a>
+     * @see <a href="https://issues.apache.org/jira/browse/IGNITE_15799">IGNITE_15799</a>
```
Comment (AMashenkov) [inline modules/core/src/test/java/org/apache/ignite/internal/testframework/WorkDirectoryExtensionTest.java:383]: Invalid link.

File: modules/client/src/test/java/org/apache/ignite/client/ClientPutGetBenchmark.java [hunk 8]
```diff
@@ -38,24 +38,24 @@
  *
  * <p>Results on i9-12900H, openjdk 11.0.18, Ubuntu 22.04:
  * Benchmark                                                Mode  Cnt      Score       Error   Units
- * ClientPutGetBenchmark.get                               thrpt    3  49418.080 ± 17639.810   ops/s (before IGNITE-18899)
```
Comment (AMashenkov) [inline modules/client/src/test/java/org/apache/ignite/client/ClientPutGetBenchmark.java:41]: Let's revert this changes.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java [hunk 9]
```diff
@@ -140,6 +139,14 @@ public static IgniteRel optimize(SqlNode sqlNode, IgnitePlanner planner) {
         }
     }
 
+    private static boolean amountOfJoinsAreBig(RelNode rel) {
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java:142]:
```suggestion
    private static boolean hasTooMuchJoins(RelNode rel) {
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/Hints.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+/**
+ * Enumeration of supported SQL hints together with help methods.
+ */
+public enum Hints {
+    //Only for test purpose!!! No any product use and not available for user
+    DISABLE_RULE(true),
+    //Disable commute joins and keeps the join order defined by the user in a query
+    ENFORCE_JOIN_ORDER,
+    EXPAND_DISTINCT_AGG;
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/Hints.java:38]:
```suggestion
    /** Only for test purpose!!! No any product use and not available for user. */
    DISABLE_RULE(true),
    /** Disable commute joins and keeps the join order defined by the user in a query. */
    ENFORCE_JOIN_ORDER,
    /** Force expanding distinct aggregates to join. */
    EXPAND_DISTINCT_AGG;
```

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java [hunk 11]
```diff
@@ -1318,4 +1344,48 @@ public Publisher<BinaryRow> lookup(int partId, HybridTimestamp timestamp, Cluste
     Predicate<SearchBounds> empty() {
         return Objects::isNull;
     }
+
+    /**
+     * Wrapper for RelOptListener with empty realization for each of their methods. Good choice in case you need implement just one or few
+     * methods for listener.
+     */
+    static class RelOptListenerWrapper implements RelOptListener {
```
Comment (AMashenkov) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java:1352]:
```suggestion
    static class NoopRelOptListener implements RelOptListener {
```

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/WorkDirectoryExtensionTest.java (hunk 7 above)
Comment (ygerzhedovich) [reply modules/core/src/test/java/org/apache/ignite/internal/testframework/WorkDirectoryExtensionTest.java:383]: this change is intentional - there is a desire to avoid linking in a code to resolved tickets because we have a bot that checks it. It looks a crutch and potentially in the future will be changed.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java (hunk 9 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java:142]: done

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/Hints.java (hunk 10 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/Hints.java:38]: thanks

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java (hunk 11 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java:1352]: ok

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/JoinCommutePlannerTest.java (hunk 6 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/JoinCommutePlannerTest.java:105]: rewrote, thanks

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java (hunk 5 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java]: ok

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/WorkDirectoryExtensionTest.java (hunk 7 above)
Comment (AMashenkov) [reply modules/core/src/test/java/org/apache/ignite/internal/testframework/WorkDirectoryExtensionTest.java:383]: I think we should either to fix the bot or remove links to resolved tickets.
Why do we need a link to a ticket if we can just write comprehensive comment instead?

**Review by lowka — APPROVED** [review]

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/WorkDirectoryExtensionTest.java (hunk 7 above)
Comment (ygerzhedovich) [reply modules/core/src/test/java/org/apache/ignite/internal/testframework/WorkDirectoryExtensionTest.java]: removed from the patch

**Review by AMashenkov — APPROVED** [review]

**Review by AMashenkov — APPROVED** [review]

**Review by AMashenkov — APPROVED** [review]
